packages/next/src/lib/metadata/resolve-metadata.tsx

```diff
@@ -0,0 +1,243 @@
+      // A number that represents which layer the item is in. Starting from 0.
+      layer: number
+      mod: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path: string
+    }
+  | {
+      type: 'icon'
+      // A number that represents which layer the item is in. Starting from 0.
+      layer: number
+      mod?: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path?: string
+    }
+
+// Merge the source metadata into the resolved target metadata.
+function merge(
+  target: ResolvedMetadata,
+  source: Metadata,
+  templateStrings: {
+    title: AbsoluteTemplateString
+    openGraphTitle: AbsoluteTemplateString
+    twitterTitle: AbsoluteTemplateString
+  }
+) {
+  let updatedStashedTitle: AbsoluteTemplateString = templateStrings.title
+  let updatedStashedOpenGraphTitle: AbsoluteTemplateString =
+    templateStrings.openGraphTitle
+  let updatedStashedTwitterTitle: AbsoluteTemplateString =
+    templateStrings.twitterTitle
+
+  for (const key_ in source) {
+    const key = key_ as keyof Metadata
+
+    switch (key) {
+      case 'other': {
+        target.other = { ...target.other, ...source.other }
+        break
+      }
+      case 'title': {
+        updatedStashedTitle = resolveTitle(templateStrings.title, source.title)
+        target.title = updatedStashedTitle
+        break
+      }
+      case 'openGraph': {
+        if (typeof source.openGraph !== 'undefined') {
+          target.openGraph = {
+            ...resolveOpenGraph(source.openGraph),
+          }
+          if (source.openGraph && 'title' in source.openGraph) {
+            updatedStashedOpenGraphTitle = resolveTitle(
+              templateStrings.openGraphTitle,
+              source.openGraph.title
+            )
+            target.openGraph.title = updatedStashedOpenGraphTitle
+          }
+        } else {
+          target.openGraph = null
+        }
+        break
+      }
+      case 'twitter': {
+        if (typeof source.twitter !== 'undefined') {
+          target.twitter = { ...source.twitter }
+          if (source.twitter && 'title' in source.twitter) {
+            updatedStashedTwitterTitle = resolveTitle(
+              templateStrings.twitterTitle,
+              source.twitter.title
+            )
+            target.twitter.title = updatedStashedTwitterTitle
+          }
+        } else {
+          target.twitter = null
+        }
+        break
+      }
+      default: {
+        // TODO: Make sure the type is correct.
+        // @ts-ignore
+        target[key] = source[key]
+        break
+      }
+    }
+  }
+
+  return {
+    title: updatedStashedTitle,
+    openGraphTitle: updatedStashedOpenGraphTitle,
+    twitterTitle: updatedStashedTwitterTitle,
+  }
```

Comment:
This function is mutative but also has a return. that's not necessarily a problem but it read confusing to me still and I think it would be much clearer if `merge` only mutated `resolvedMetadata`

I'd suggest we stash the current template from template strings on `resolvedMetadata` at the start of the for loop if we entered a new layer. then pass them into this merge function and use them to mutate the template strings in resolvedMetadata.

The return would not require an object allocation, there would be less total copying of stashed titles, and the logic for when to "commit" a title goes away